### PR TITLE
Add end-2-end tests target for different runtimes

### LINKED ISSUE #104: Need to add End2End integration test in CI system
Need to verify that Lighthouse can start up correctly, inside a Docker container with the various runtimes we support.

### PR DESCRIPTION
**Pull request by IgorFedchenko**
Close #104

### CODE REVIEW

**Comment by IgorFedchenko** [comment]
Right know only docker containers are checked - so in draft state until fill .NET Framework check support will be added

**Comment by Aaronontheweb** [comment]
> Right know only docker containers are checked - so in draft state until fill .NET Framework check support will be added

That makes sense - we only need this to really run on one platform. No need to do both .NET Framework and .NET Core.

**Comment by IgorFedchenko** [comment]
@Aaronontheweb 
> No need to do both .NET Framework and .NET Core.

Good to know. Then will make tests pass and will switch PR to ready-for-review state.
Seems like windows docker container does not want to start under windows.

File: src/Lighthouse/LighthouseHostFactory.cs [hunk 1]
```diff
@@ -45,7 +45,7 @@ public static class LighthouseHostFactory
             if (lighthouseConfig != null && IsNullOrEmpty(systemName))
                 systemName = lighthouseConfig.GetString("actorsystem", systemName);
 
-            ipAddress = clusterConfig.GetString("akka.remote.dot-netty.tcp.public-hostname");
+            ipAddress = clusterConfig.GetString("akka.remote.dot-netty.tcp.public-hostname", "127.0.0.1");
```
Comment (IgorFedchenko) [inline src/Lighthouse/LighthouseHostFactory.cs]: When trying to start Lighthouse without environment configuration provided, it does not find `public-hostname` setting in hocon and throws NRE later (when trying to make `ipAddress.Trim()`).
I think warnings about issues connecting to `127.0.0.1` is more informative, and at least does not look so scary :) Maybe even better would be to add some `log.warning` about missing public hostname here?

File: build.fsx [hunk 2]
```diff
@@ -128,6 +130,76 @@ Target "NBench" <| fun _ ->
...
+                
+    let connectLighthouse () =
+        printfn "Connecting Lighthouse..."
+        try
+            ExecProcess(fun info -> 
+                info.FileName <- "pbm") (System.TimeSpan.FromSeconds LighthouseConnectTimeout) |> ignore
+            // If process returned, this means that pbm failed to connect
+            raise (ConnectionFailure "Failed to connect Lighthouse from pbm")
+        with
+            | ConnectionFailure(str) -> reraise()
```
Comment (IgorFedchenko) [inline build.fsx]: Unfortunately, `ExecProcess` throws `System.Exception` on process timeout, so have to `reraise` custom exception here just to fail for different reason.

File: build.fsx [hunk 3]
```diff
@@ -128,6 +130,76 @@ Target "NBench" <| fun _ ->
...
+    
+    installPbm()
+    startLighthouseDocker dockerFileForTest
+    try       
+        connectLighthouse()
+    finally
+        stopLighthouseDocker dockerFileForTest
+        
+    // Test Full .NET Framework version under windows only
+    // TODO: To make this work, need to start lighthouse and pbm as two parallel processes
```
Comment (IgorFedchenko) [inline build.fsx]: Left this here for any future work, if required.

**Comment by IgorFedchenko** [comment]
@Aaronontheweb 
After container build:
```
docker build -f Dockerfile-windows -t lighthouse:1.2.0 -t lighthouse:latest
```
It is failing to make this:
```
docker run <options> lighthouse:latest
```
With "The operating system of the container does not match the operating system of the host" error. No idea why - will try to make some research for that.

**Comment by IgorFedchenko** [comment]
So, this may be related with the difference between Windows version used in CI and `mcr.microsoft.com/dotnet/core/sdk:2.1-nanoserver-1803` container embedded version.

In this case, we need to either change CI Windows version, or switch to another base container. But in the second case, change may introduce conflicts for people that are using Lighthouse windows container in their environment.

File: src/Lighthouse/LighthouseHostFactory.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/Lighthouse/LighthouseHostFactory.cs]: I think that's a good idea - we should add it.

File: build.fsx (hunk 2 above)
Comment (Aaronontheweb) [reply build.fsx]: LGTM

File: build.fsx (hunk 3 above)
Comment (Aaronontheweb) [reply build.fsx]: We can skip doing this on .NET Framework for the time being - the behavior differences are minimal.

**Comment by Aaronontheweb** [comment]
> In this case, we need to either change CI Windows version, or switch to another base container. But in the second case, change may introduce conflicts for people that are using Lighthouse windows container in their environment.

@IgorFedchenko let's change the base container image and give that a try.

**Comment by IgorFedchenko** [comment]
@Aaronontheweb 
> @IgorFedchenko let's change the base container image and give that a try.

Which version to switch to? There are few of them [here](https://hub.docker.com/_/microsoft-dotnet-core-sdk) in the bottom. Or I could just try one after another... What Windows version are we using? Maybe this will help me to find appropriate version faster.

**Comment by Aaronontheweb** [comment]
@IgorFedchenko why don't we do this first actually - let's change the `.yaml` files for Azure DevOps to run on Windows 2019 instead - that might be part of the issue.

**Comment by IgorFedchenko** [comment]
@Aaronontheweb `.yaml` files are already using `vmImage: windows-2019` setting, but this may mean that we should update base image version, indeed.
`windows-2019` image content [description](https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2019-Server2019-Readme.md) says that is uses `WinAppDriver
Version: 1.1.1809.18001` - so who knows, maybe this `1809` number is correlating with `2.1-nanoserver-1809` image we can use... I will try it right now.

**Comment by IgorFedchenko** [comment]
Nope, did not help...
There are actually only two more images with newer version - let's make a quick try with them.

**Comment by Aaronontheweb** [comment]
they made some big changes to 2019 - should have the ability to execute Linux images natively now. I wonder if that's the issue - that Windows Server 2019 defaults to running Linux ones out of the box.

**Comment by IgorFedchenko** [comment]
When trying to build docker image with `1903` or `1909` version, we are now getting error:
```
a Windows version 10.0.18362-based image is incompatible with a 10.0.17763 host
```

So seems like this images are too new for this host...

**Comment by IgorFedchenko** [comment]
> they made some big changes to 2019 - should have the ability to execute Linux images natively now. I wonder if that's the issue - that Windows Server 2019 defaults to running Linux ones out of the box.

That's an interesting idea.

**Comment by IgorFedchenko** [comment]
@Aaronontheweb [Here](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/container-phases?view=azure-devops&tabs=yaml#windows-containers) they say that no problem should be there

**Comment by IgorFedchenko** [comment]
@Aaronontheweb Ah, and then they say that you should run `win1803` version to run windows containers... Let me try it now.

**Comment by IgorFedchenko** [comment]
New windows version should be able to run windows containers, but it does not seem to support Full .NET Framework and thus refuses to build project with multi-targeting.
Will make it build under `netcoreapp2.1` only (because this is the one we are running in Windows container)

**Comment by IgorFedchenko** [comment]
All right... So since current windows PR check calls `build.cmd all` target, it is trying to build under full .NET framework and fails on new `win1803` version.
So had to move `RunTestsOnRuntimes` target from `All` target, and place them into separate Azure Pipelines jobs - 'All` is running on `windows-2019` as before, and `RunTestsOnRuntimes` is running on `win1803`.

**Comment by IgorFedchenko** [comment]
Mm, I see, I did not understood right how FAKE is interpreting target dependencies... So it takes all defined references in one single graph - well, this is something I could expect.

All right, I will create some separate targets to avoid building under full .net framework (just need to skip `Build` step and start with publishing and building docker container).

**Comment by IgorFedchenko** [comment]
Yey :)

**Comment by Aaronontheweb** [comment]
Nice work

**Pull request merged**